File: src/main/java/org/artembogomolova/demo/webapp/config/SecurityConfig.java
package org.artembogomolova.demo.webapp.config;

import lombok.RequiredArgsConstructor;
import org.artembogomolova.demo.webapp.security.authrorize.request.SecurityExpressionHandlerImpl;
import org.springframework.context.annotation.Bean;
import org.springframework.security.access.AccessDecisionManager;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.security.web.authentication.AuthenticationFailureHandler;
import org.springframework.security.web.authentication.AuthenticationSuccessHandler;
import org.springframework.security.web.authentication.logout.LogoutHandler;
import org.springframework.security.web.authentication.logout.LogoutSuccessHandler;

@EnableWebSecurity
@RequiredArgsConstructor
public class SecurityConfig extends WebSecurityConfigurerAdapter {

    public static final String LOGIN_URL = "/login";
    private static final String LOGOUT_URL = "/logout";
    private static final String[] UNSECURED_RESOURCES = {"/webjars/**", "/css/**", "/resources/**"};
    private static final String[] SECURED_RESOURCES = {"/api/**"};

    private final AuthenticationSuccessHandler loginSuccessfulHandler;
    private final AuthenticationFailureHandler loginFailureHandler;
    private final LogoutSuccessHandler logoutSuccessHandler;
    private final LogoutHandler logoutHandler;
    private final UserDetailsService userDetailsService;
    private final AccessDecisionManager accessDecisionManager;
    private final SecurityExpressionHandlerImpl securityExpressionHandlerImpl;


    @Bean
    public PasswordEncoder getPasswordEncoder() {
        return new BCryptPasswordEncoder();
    }

    @Override
    protected void configure(HttpSecurity http)
            throws
            Exception {
        http.authorizeRequests()
                .accessDecisionManager(accessDecisionManager)
                .expressionHandler(securityExpressionHandlerImpl)
                .filterSecurityInterceptorOncePerRequest(true)
                /* api methods need authorization*/
                .antMatchers(SECURED_RESOURCES)
                .authenticated()
                /*but unsecured resources(css, static resources) don't need*/
                .antMatchers(UNSECURED_RESOURCES)
                .anonymous()
                .and()
                /*login page with failure callback*/
                .formLogin()
                .loginPage(LOGIN_URL)
                .permitAll()
                .successHandler(loginSuccessfulHandler)
                .failureHandler(loginFailureHandler)
                .and()
                /*logout page*/
                .logout()
                .permitAll()
                .and()
                .logout()
                .logoutUrl(LOGOUT_URL)
                .logoutSuccessUrl(LOGIN_URL)
                .logoutSuccessHandler(logoutSuccessHandler)
                .invalidateHttpSession(true)
                .addLogoutHandler(logoutHandler)
                .and()
                /*user service with detail data about logged user*/
                .userDetailsService(userDetailsService)
                .exceptionHandling();
    }

}

File: app/src/main/kotlin/org/artembogomolova/demo/webapp/main/config/SecurityConfig.kt
package org.artembogomolova.demo.webapp.main.config

import org.artembogomolova.demo.webapp.main.controller.PublicController
import org.artembogomolova.demo.webapp.main.domain.auth.PredefinedUserRole
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.ComponentScan
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configuration.WebSecurityConfigurerAdapter
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.authentication.AnonymousAuthenticationFilter
import org.springframework.security.web.authentication.AuthenticationFailureHandler
import org.springframework.security.web.authentication.AuthenticationSuccessHandler
import org.springframework.security.web.authentication.logout.LogoutHandler
import org.springframework.security.web.authentication.logout.LogoutSuccessHandler

@Configuration
@EnableWebSecurity
@ComponentScan(
    value = ["org.artembogomolova.demo.webapp.main.security",
        "org.artembogomolova.demo.webapp.main.controller",
        "org.artembogomolova.demo.webapp.main.filter"]
)
class SecurityConfig : WebSecurityConfigurerAdapter() {

    @Autowired
    lateinit var loginSuccessfulHandler: AuthenticationSuccessHandler

    @Autowired
    lateinit var loginFailureHandler: AuthenticationFailureHandler

    @Autowired
    lateinit var logoutSuccessHandler: LogoutSuccessHandler

    @Autowired
    lateinit var logoutHandler: LogoutHandler

    //    private val userDetailsService: UserDetailsService? = null,
    @Autowired
    lateinit var anonymousAuthenticationFilterImpl: AnonymousAuthenticationFilter

    @get:Bean
    val passwordEncoder: PasswordEncoder
        get() = BCryptPasswordEncoder()

    @Throws(Exception::class)
    override fun configure(http: HttpSecurity) {
        configureLoginProcess(http)
        configureLogoutProcess(http)
        configurePublicResources(http)
    }

    @Throws(Exception::class)
    private fun configurePublicResources(http: HttpSecurity) {
        http.authorizeRequests()
            .antMatchers(*PublicController.unsecuredResources)
            .hasAnyAuthority(*PredefinedUserRole.GUEST.privilegesAsArray)
        /*For guest main page*/http.addFilterAt(anonymousAuthenticationFilterImpl, AnonymousAuthenticationFilter::class.java)
    }

    @Throws(Exception::class)
    private fun configureLogoutProcess(http: HttpSecurity) {
        http.logout()
            .logoutUrl(LOGOUT_URL)
            .logoutSuccessUrl(PublicController.LOGIN_URL)
            .logoutSuccessHandler(logoutSuccessHandler)
            .invalidateHttpSession(true)
            .addLogoutHandler(logoutHandler)
    }

    /**
     * login page with failure callback
     */
    @Throws(Exception::class)
    private fun configureLoginProcess(http: HttpSecurity) {
        http.formLogin()
            .loginPage(PublicController.LOGIN_URL)
            .successHandler(loginSuccessfulHandler)
            .failureHandler(loginFailureHandler)
    }

    companion object {
        private const val LOGOUT_URL = "/logout"
        private val SECURED_RESOURCES = arrayOf("/api/**")
    }
}